File: js-template/leetcode/editor/cn/maximum-subarray.js
/*
 * @lc app=leetcode.cn id=53 lang=javascript
 * @lcpr version=30404
 *
 * [53] 最大子数组和
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    if (nums.length === 1)
        return nums[0];

    var sums = [];
    var min, max;

    for (let i = 0; i < nums.length; i++) {
        if (sums.length === 0) {
            sums.push(nums[i]);
            min = max = nums[i];
        } else {
            const data = sums[i-1] + nums[i];
            let temp = data - Math.min(min, 0);
            if (temp > max)
                max = temp;
            if (data < min)
                min = data;
            
            sums.push(sums[i-1] + nums[i]);
        }
    }

    return max;
};
// @lc code=end

// your test code here


console.log(maxSubArray([-2, -1]))


/*
// @lcpr case=start
// [-2,1,-3,4,-1,2,1,-5,4]\n
// @lcpr case=end

// @lcpr case=start
// [1]\n
// @lcpr case=end

// @lcpr case=start
// [5,4,-1,7,8]\n
// @lcpr case=end

// @lcpr case=start
// [-2, -1]\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/maximum-subarray.cpp
/*
 * @lc app=leetcode.cn id=53 lang=cpp
 * @lcpr version=30404
 *
 * [53] 最大子数组和
 */

#include <iostream>
#include <vector>
#include <string>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int result = nums[0];
        int minSum = 0;
        int sum = 0;
        for (auto &&num : nums) {
            sum += num;
            result = max(result, sum - minSum);
            if (sum < minSum)
                minSum = sum;
        }
        return result;
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
    // vector<int> nums = {5,4,-1,7,8};
    vector<int> nums = {-1};
    solution.maxSubArray(nums);
}



/*
// @lcpr case=start
// [-2,1,-3,4,-1,2,1,-5,4]\n
// @lcpr case=end

// @lcpr case=start
// [1]\n
// @lcpr case=end

// @lcpr case=start
// [5,4,-1,7,8]\n
// @lcpr case=end

 */

